src/kubernetes_multus.py

```diff
@@ -0,0 +1,276 @@
+            logger.info("No network annotations were provided")
+            return
+        statefulset = self.client.get(res=StatefulSet, name=name, namespace=self.namespace)
+        statefulset.spec.template.metadata.annotations["k8s.v1.cni.cncf.io/networks"] = json.dumps(  # type: ignore[attr-defined]  # noqa: E501
+            [network_annotation.dict() for network_annotation in network_annotations]
+        )
+
+        # statefulset.spec.template.spec.containers[2].securityContext.privileged = True
+        # statefulset.spec.template.spec.containers[2].securityContext.capabilities = Capabilities(
+        #     add=[
+        #         "NET_ADMIN",
+        #     ]
+        # )
+
+        self.client.patch(
+            res=StatefulSet,
+            name=name,
+            obj=statefulset,
+            patch_type=PatchType.MERGE,
+            namespace=self.namespace,
+        )
+        logger.info(f"Multus annotation added to {name} Statefulset")
+
+    def statefulset_is_patched(
+        self, name: str, network_annotations: List[NetworkAnnotation]
+    ) -> bool:
+        """Returns whether the statefulset has the expected multus annotation.
+
+        Args:
+            name: Statefulset name.
+            network_annotations: List of network annotations
+
+        Returns:
+            bool: Whether the statefulset has the expected multus annotation.
+        """
+        statefulset = self.client.get(res=StatefulSet, name=name, namespace=self.namespace)
+        if "k8s.v1.cni.cncf.io/networks" not in statefulset.spec.template.metadata.annotations:  # type: ignore[attr-defined]  # noqa: E501
+            logger.info("Multus annotation not yet added to statefulset")
+            return False
+        if json.loads(
+            statefulset.spec.template.metadata.annotations["k8s.v1.cni.cncf.io/networks"]  # type: ignore[attr-defined]  # noqa: E501
+        ) != [network_annotation.dict() for network_annotation in network_annotations]:
+            logger.info("Existing annotation are not identical to the expected ones")
+            return False
+        logger.info("Multus annotation already added to statefulset")
+        return True
+
+
+class KubernetesMultusCharmLib(Object):
+    """Class to be instantiated by charms requiring Multus networking."""
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        network_attachment_definitions: List[NetworkAttachmentDefinition],
+        network_annotations: List[NetworkAnnotation],
+    ):
+        super().__init__(charm, "kubernetes-multus")
+        self.network_attachment_definitions = network_attachment_definitions
+        self.network_annotations = network_annotations
+        self.framework.observe(charm.on.install, self._patch)
+        self.framework.observe(charm.on.upgrade_charm, self._patch)
+        self.framework.observe(charm.on.remove, self._on_remove)
+
+    def _patch(self, event: Union[InstallEvent, UpgradeCharmEvent]) -> None:
```

Comment:
```suggestion
    def _patch(self, event: ops.framework.EventBase) -> None:
```
We are not using anything on the event, so using type `EventBase` would be better here.

Reply:
Done